### ISSUE TITLE
Count: removing an item from a list forces unexpected changes

### ISSUE THREAD

**Issue by rohrerb** [hashicorp/terraform#18349]
This is in reference to testing the following PR - https://github.com/terraform-providers/terraform-provider-azurerm/pull/1207 (Working great so far!  Thank you @tombuildsstuff )

The goal would be to build out an environment with more than 2 VMs with X data disks.  It works perfectly using this TF file - https://gist.github.com/rohrerb/0cac6f3b24691674538067fd491c460d

**Builds out the following:**
https://imgur.com/hS8bb1X 

Once the environment is built, increment the variable for data_disk_count from "1" to "2" and re-run a plan.  

> variable "data_disk_count" {  default = 2}

You will see that it wants to destroy the the 2nd VM disk because the element changed based on the new count.index.   I know that AWS has had attached disk logic for a while and im hoping that someone figured out how to apply this sort of logic.

>> @tombuildsstuff  asked me to move to this repo (https://github.com/terraform-providers/terraform-provider-azurerm/issues/1441)

**Comment by tombuildsstuff** [comment]
Re-tagging this as `core` since I believe this is a core issue (removing an element from the middle of a list shouldn't destroy existing elements)

**Comment by apparentlymart** [comment]
Hi @rohrerb! Sorry for this annoying behavior.

The intended solution to this problem is being discussed over in #17179. For this particular issue we're not going to be be able to get it all done for the next major release but we _are_ laying the groundwork that should hopefully make this feature easier to implement (supporting strings as resource instance keys, etc.)

In the mean time, I expect the problem here is the introduction of a new `<computed>` value into one of the "splat" expression results (like `azurerm_managed_disk.data_disk.*.id`), which currently causes Terraform to see the whole result as unknown. There are a few common workarounds for this that may work in your case, but there are enough moving parts in your configuration that I'm not 100% sure...

* Instead of using `element(list, index)`, use `list[index]`. The native index syntax allows Terraform core to understand that you're only accessing one index, and so other elements of the list being unknown won't cause the whole result to be unknown. (This particular part of the issue will be fixed in the next major release, to make `element(list, index)` behave the same as `list[index]` when the index is within the list bounds).
* When applying a change like this for the first time, use the `-target` argument to gradually add the new components while leaving the existing components untouched. Once this is complete, the normal un-targeted apply should then work as expected again, because there are no unknown ids in the lists.

  ```
  # First, create just the new disk
  terraform apply -target=azurerm_managed_disk.data_disk[1]

  # Then, attach the new disk to the instance
  terraform apply -target=azurerm_virtual_machine_data_disk_attachment.data_disk_attach[1]

  # Verify that no further changes are required
  terraform apply
  ```

Some parts of this problem will be addressed in [the next major release](https://www.hashicorp.com/blog/terraform-0-1-2-preview), so you may find that these changes are sufficient to make your configuration here work even without #17179 being implemented. We'll check in here again closer to the release to check if the changes have fixed this issue or if there's a better workaround available as a result of these changes.

**Comment by apparentlymart** [comment]
Hi again, @rohrerb! Sorry for the long silence here.

Now that we've got the changes for the forthcoming v0.12.0 release merged down into master, I wrote the following simpler config to see where we stand here:

```hcl
variable "instance_count" {
  default = 1
}

resource "null_resource" "a" {
  count = var.instance_count
}

resource "null_resource" "b" {
  count = var.instance_count

  triggers = {
    upstream = null_resource.a.*.id[count.index]
  }
}
```

In v0.11 this hits the same bug you encountered where increasing count after the zero-index instances of both resources exist wants to update or replace the existing `null_resource.b[0]` along with creating `null_resource.a[1]` and `null_resource.b[1]`.

I have verified with the v0.12.0-alpha2 build that _this_ part of the problem is now resolved:

```
$ terraform apply

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # null_resource.a[0] will be created
  + resource "null_resource" "a" {
      + id = (known after apply)
    }

  # null_resource.b[0] will be created
  + resource "null_resource" "b" {
      + id       = (known after apply)
      + triggers = (known after apply)
    }

Plan: 2 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

null_resource.a[0]: Creating...
null_resource.a[0]: Creation complete after 0s [id=1373382676816235508]
null_resource.b[0]: Creating...
null_resource.b[0]: Creation complete after 0s [id=2335790106368102365]

Apply complete! Resources: 2 added, 0 changed, 0 destroyed.
```

```
$ terraform apply -var="instance_count=2"
null_resource.a[0]: Refreshing state... [id=1373382676816235508]
null_resource.b[0]: Refreshing state... [id=2335790106368102365]

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # null_resource.a[1] will be created
  + resource "null_resource" "a" {
      + id = (known after apply)
    }

  # null_resource.b[1] will be created
  + resource "null_resource" "b" {
      + id       = (known after apply)
      + triggers = (known after apply)
    }

Plan: 2 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

null_resource.a[1]: Creating...
null_resource.a[1]: Creation complete after 0s [id=1756803054747501777]
null_resource.b[1]: Creating...
null_resource.b[1]: Creation complete after 0s [id=5281863472455805767]

Apply complete! Resources: 2 added, 0 changed, 0 destroyed.
```

This works now because adding a new unknown id to the end of the list produced by `null_resource.a.*.id` doesn't cause `null_resource.a.*.id[0]` to return unknown anymore. The unknown-ness is limited only to `null_resource.a.*.id[1]`.

Likewise I can reduce back down to just one instance without also replacing the zeroth instance:

```
$ terraform apply
null_resource.a[1]: Refreshing state... [id=1756803054747501777]
null_resource.a[0]: Refreshing state... [id=1373382676816235508]
null_resource.b[1]: Refreshing state... [id=5281863472455805767]
null_resource.b[0]: Refreshing state... [id=2335790106368102365]

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  - destroy

Terraform will perform the following actions:

  # null_resource.a[1] will be destroyed
  - resource "null_resource" "a" {
      - id = "1756803054747501777" -> null
    }

  # null_resource.b[1] will be destroyed
  - resource "null_resource" "b" {
      - id       = "5281863472455805767" -> null
      - triggers = {
          - "upstream" = "1756803054747501777"
        } -> null
    }

Plan: 0 to add, 0 to change, 2 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

null_resource.b[1]: Destroying... [id=5281863472455805767]
null_resource.b[1]: Destruction complete after 0s
null_resource.a[1]: Destroying... [id=1756803054747501777]
null_resource.a[1]: Destruction complete after 0s

Apply complete! Resources: 0 added, 2 changed, 0 destroyed.
```

This all works as expected because in these cases we are working entirely with counts and not trying to correlate instances with a list provided separately. #17179 will be the answer to the problem of making resource instances that correlate with elements from a collection, where using a map will give a way to use string keys rather than list positions as the identifiers for the instances.

Since the increment/decrement issue is resolved in master and we already have #17179 covering the `for_each` feature, I'm going to close this out now. The fix will be included in the forthcoming v0.12.0 final release. Thanks for reporting this, and sorry again for the delay in getting this fixed.
